### PR TITLE
feat: Added MSSQL

### PR DESCRIPTION
**Pull request by cliedeman**
First pass at MSSQL support

Outstanding Issues

- [ ] Fix CI
- [ ] MSSQL Unique constraints do not allow duplicate nulls. This causes a lot of unrelated tests to fail
- [ ] MSSQL Is more strict about [cascading deletes](https://stackoverflow.com/questions/851625/foreign-key-constraint-may-cause-cycles-or-multiple-cascade-paths/852047#852047)

### CODE REVIEW

File: dialect/sql/scan.go [hunk 1]
```diff
@@ -29,6 +30,10 @@ func ScanOne(rows ColumnScanner, v interface{}) error {
 		return fmt.Errorf("sql/scan: unexpected number of columns: %d", n)
 	}
 	if !rows.Next() {
+		if rows.Err() != nil {
```
Comment (cliedeman) [inline dialect/sql/scan.go:33]: This is an extremely odd case. The error on rows is only present after the call to `Next`

**Comment by cliedeman** [comment]
Ok time for some more detailed notes.

So a MSSQL driver is mostly implementable with the current interfaces so thats good. Most of the issues stem from differences in MSSQL. 

 ### [Cascading Deletes](https://stackoverflow.com/questions/851625/foreign-key-constraint-may-cause-cycles-or-multiple-cascade-paths/852047#852047)

Example 1
```
CREATE TABLE [nodes]
(
    [id]        bigint IDENTITY (1,1) NOT NULL,
    [value]     bigint                NULL,
    [node_next] bigint UNIQUE         NULL,
    CONSTRAINT [nodes_id_pk] PRIMARY KEY ([id])
)

ALTER TABLE [nodes]
    ADD CONSTRAINT [nodes_nodes_next] FOREIGN KEY ([node_next]) REFERENCES [nodes] ([id]) ON DELETE SET NULL;
```

Error:
```
[2020-05-08 14:06:45] [S0000][1785] Introducing FOREIGN KEY constraint 'nodes_nodes_next' on table 'nodes' may cause cycles or multiple cascade paths. Specify ON DELETE NO ACTION or ON UPDATE NO ACTION, or modify other FOREIGN KEY constraints.
```

Resolution: 
Change the constraint to 
```
ALTER TABLE [nodes]
    ADD CONSTRAINT [nodes_nodes_next] FOREIGN KEY ([node_next]) REFERENCES [nodes] ([id]) ON DELETE NO ACTION ;
```

This requires  overriding the cascade rule for Edges

Example 2
```
CREATE TABLE [users]
(
    [id]            bigint IDENTITY (1,1) NOT NULL,
    [optional_int]  bigint                NULL,
    [age]           bigint                NOT NULL,
    [name]          varchar(255)          NOT NULL,
    [last]          varchar(255)          NOT NULL DEFAULT 'unknown',
    [nickname]      varchar(255) UNIQUE   NULL,
    [phone]         varchar(255) UNIQUE   NULL,
    [password]      varchar(255)          NULL,
    [role]          varchar(255)          NOT NULL,
    [sso_cert]      varchar(255)          NULL,
    [group_blocked] bigint                NULL,
    [user_spouse]   bigint UNIQUE         NULL,
    [user_parent]   bigint                NULL,
    CONSTRAINT [users_id_pk] PRIMARY KEY ([id])
)

CREATE TABLE [pets]
(
    [id]        bigint IDENTITY (1,1) NOT NULL,
    [name]      varchar(255)          NOT NULL,
    [user_pets] bigint                NULL,
    [user_team] bigint UNIQUE         NULL,
    CONSTRAINT [pets_id_pk] PRIMARY KEY ([id])
)

ALTER TABLE [pets]
    ADD CONSTRAINT [pets_users_pets] FOREIGN KEY ([user_pets]) REFERENCES [users] ([id]) ON DELETE SET NULL
ALTER TABLE [pets]
    ADD CONSTRAINT [pets_users_team] FOREIGN KEY ([user_team]) REFERENCES [users] ([id]) ON DELETE SET NULL

```

Error 
```
[2020-05-08 14:22:13] [S0000][1785] Introducing FOREIGN KEY constraint 'pets_users_team' on table 'pets' may cause cycles or multiple cascade paths. Specify ON DELETE NO ACTION or ON UPDATE NO ACTION, or modify other FOREIGN KEY constraints.
```

Here the first constraint passes but the second constraint fails

### [Nulls are not Unique]()

Example
```
INSERT INTO [users] ([age], [name], [last], [nickname], [role]) OUTPUT INSERTED.[id]  VALUES (@p1, @p2, @p3, @p4, @p5) 
args=[1 foo unknown baz user]

INSERT INTO [users] ([age], [name], [last], [nickname], [role]) OUTPUT INSERTED.[id]  VALUES (@p1, @p2, @p3, @p4, @p5) 
args=[1 bar unknown baz user]
```

Error:
```
insert node to table "users": mssql: Violation of UNIQUE KEY constraint 'UQ__users__AFA785105E5B2240'. Cannot insert duplicate key in object 'dbo.users'. The duplicate key value is (<NULL>). [recovered]
```

This actually fails with the wrong error.
Here the insert is failing because phone number is null.
This is quite problematic because lots of the test cases have nulls in them that do not cause issues on the other RDBMS

Resolution
Option 1
Implement filtered indexes
Eg

Instead of 
```
CREATE TABLE [users]
(
    [id]            bigint IDENTITY (1,1) NOT NULL,
    [optional_int]  bigint                NULL,
    [age]           bigint                NOT NULL,
    [name]          varchar(255)          NOT NULL,
    [last]          varchar(255)          NOT NULL DEFAULT 'unknown',
    [nickname]      varchar(255) UNIQUE   NULL,
    [phone]         varchar(255) UNIQUE   NULL,
    [password]      varchar(255)          NULL,
    [role]          varchar(255)          NOT NULL,
    [sso_cert]      varchar(255)          NULL,
    [group_blocked] bigint                NULL,
    [user_spouse]   bigint UNIQUE         NULL,
    [user_parent]   bigint                NULL,
    CONSTRAINT [users_id_pk] PRIMARY KEY ([id])
)
```
do

```
CREATE TABLE [users]
(
    [id]            bigint IDENTITY (1,1) NOT NULL,
    [optional_int]  bigint                NULL,
    [age]           bigint                NOT NULL,
    [name]          varchar(255)          NOT NULL,
    [last]          varchar(255)          NOT NULL DEFAULT 'unknown',
    [nickname]      varchar(255) UNIQUE   NULL,
    [phone]         varchar(255) UNIQUE   NULL,
    [password]      varchar(255)          NULL,
    [role]          varchar(255)          NOT NULL,
    [sso_cert]      varchar(255)          NULL,
    [group_blocked] bigint                NULL,
    [user_spouse]   bigint UNIQUE         NULL,
    [user_parent]   bigint                NULL,
    CONSTRAINT [users_id_pk] PRIMARY KEY ([id])
)
create unique index [users_nickname] on [users] ([nickname]) where [nickname] is not null
create unique index [users_phone] on [users] ([nickname]) where [phone] is not null
```

This will allow nulls to not cause duplicate errors. 

### Pagination requires Order By

Example

```
SELECT DISTINCT [users].[id],
                [users].[optional_int],
                [users].[age],
                [users].[name],
                [users].[last],
                [users].[nickname],
                [users].[phone],
                [users].[password],
                [users].[role],
                [users].[sso_cert]
FROM [users] OFFSET @p1 ROWS
FETCH NEXT @p2 ROWS ONLY 
```

Error:
```
Invalid usage of the option NEXT in the FETCH statement.
```

Resolution:
Only use pagination with order by

### Can't drop circular references

Example Error
```
he DELETE statement conflicted with the REFERENCE constraint "user_friends_user_id". The conflict occurred in database "master", table "dbo.user_friends", column 'user_id'. [recovered]
```

This requires deleting the parent at the same time as the child.

**Comment by ProximaB** [comment]
@cliedeman ✋ Do you plan to continue to develop this? May I help you somehow? 🙂

**Comment by cliedeman** [comment]
Hi @ProximaB 

I don't plan on continuing this PR but I am an interested in some of the features. Like filtered indexes which also applies to Postgres

**Pull request closed without merging**